Support M1 MacBooks darwin/arm64 on the client-side

dims opened the pull request:
Please test using:
```
build/run.sh make generated_files && make quick-release-images
```

Signed-off-by: Davanum Srinivas <davanum@gmail.com>

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**

/kind feature

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind design

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #97550

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Add support to generate client-side binaries for new darwin/arm64 platform
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

dims commented:
/sig release
/area release-eng
/priority important-soon

@kubernetes/release-managers WDYT? what else do we need?

dims commented:
/hold

hasheddan commented:
@dims we are going to be working on formalizing the process for supported platforms in the 1.21 cycle (https://github.com/kubernetes/sig-release/issues/1337). This, along with riscv, would be a good candidate to be the pilots for this process. What do you think?

dims commented:
@hasheddan +1 sounds good!

hwdef commented:
/lgtm

dims commented:
we'll need to wait for golang 1.16 ... forgot to mention :) 
https://github.com/golang/go/blob/master/doc/go1.16.html#L36-L43

dims commented:
/lgtm cancel

dims committed:
Support M1 MacBooks darwin/arm64 on the client-side

testing using:
```
build/run.sh make generated_files && make quick-release-images
```

Signed-off-by: Davanum Srinivas <davanum@gmail.com>

slayerjain commented:
@dims golang 1.16 is released! :)

dims commented:
@slayerjain this needs to merge first https://github.com/kubernetes/kubernetes/pull/98572

sharadjain21 commented:
Hi @dims / Team

Could you please suggest a way of installation of kubectl on new Apple Mac M1. 
I tried using below commands and seems failing. 
 
_om@Oms-MacBook-Pro ~ % brew install kubectlError: kubernetes-cli: no bottle available!You can try to install from source with:  brew install --build-from-source kubernetes-cliPlease note building from source is unsupported. You will encounter buildfailures with some formulae. If you experience any issues please create pullrequests instead of asking for help on Homebrew's GitHub, Twitter or any otherofficial channels.om@Oms-MacBook-Pro ~ % brew install kubernetes-cliError: kubernetes-cli: no bottle available!You can try to install from source with:  brew install --build-from-source kubernetes-cliPlease note building from source is unsupported. You will encounter buildfailures with some formulae. If you experience any issues please create pullrequests instead of asking for help on Homebrew's GitHub, Twitter or any otherofficial channels.om@Oms-MacBook-Pro ~ % brew install --build-from-source kubernetes-cli==> Downloading https://homebrew.bintray.com/bottles/go-1.16.arm64_big_sur.bottle.tar.gz
curl: (35) error:1400410B:SSL routines:CONNECT_CR_SRVR_HELLO:wrong version numberError: Failed to download resource "go"Download failed: https://homebrew.bintray.com/bottles/go-1.16.arm64_big_sur.bottle.tar.gz_

Could you please suggest.

Rotzke commented:
> Hi @dims / Team
> 
> Could you please suggest a way of installation of kubectl on new Apple Mac M1. 
> I tried using below commands and seems failing.
>  
> _om@Oms-MacBook-Pro ~ % brew install kubectlError: kubernetes-cli: no bottle available!You can try to install from source with:  brew install --build-from-source kubernetes-cliPlease note building from source is unsupported. You will encounter buildfailures with some formulae. If you experience any issues please create pullrequests instead of asking for help on Homebrew's GitHub, Twitter or any otherofficial channels.om@Oms-MacBook-Pro ~ % brew install kubernetes-cliError: kubernetes-cli: no bottle available!You can try to install from source with:  brew install --build-from-source kubernetes-cliPlease note building from source is unsupported. You will encounter buildfailures with some formulae. If you experience any issues please create pullrequests instead of asking for help on Homebrew's GitHub, Twitter or any otherofficial channels.om@Oms-MacBook-Pro ~ % brew install --build-from-source kubernetes-cli==> Downloading https://homebrew.bintray.com/bottles/go-1.16.arm64_big_sur.bottle.tar.gz curl: (35) error:1400410B:SSL routines:CONNECT_CR_SRVR_HELLO:wrong version numberError: Failed to download resource "go"Download failed: https://homebrew.bintray.com/bottles/go-1.16.arm64_big_sur.bottle.tar.gz_
> 
> Could you please suggest.

You don't really need brew for this, I am temporary using amd64 binary with Rosetta 2, try that.

dims commented:
@sharadjain21 - do you want to install go 1.16 arm64 on your mac m1 and try a simple `go install k8s.io/kubernetes/cmd/kubectl`?

sharadjain21 commented:
@dims
yeah actually going to learn k8s and first steps were of installation and
there i got stuck.. could you please suggest if there is something i can
try please.

Thanks
Sharad

On Mon, Mar 1, 2021 at 12:29 AM Davanum Srinivas <notifications@github.com>
wrote:

> @sharadjain21 <https://github.com/sharadjain21> - do you want to install
> go 1.16 arm64 on your mac m1 and try a simple go install
> k8s.io/kubernetes/cmd/kubectl?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/97743#issuecomment-787502921>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AKACI4KT6FTJHRIDAPXA4M3TBKHCHANCNFSM4VWVF6PQ>
> .
>

Rotzke commented:
> @sharadjain21 - do you want to install go 1.16 arm64 on your mac m1 and try a simple `go install k8s.io/kubernetes/cmd/kubectl`?

Why not use binary?

sharadjain21 commented:
you mean to clone k8s repo and then make it .. right?

On Mon, Mar 1, 2021 at 12:38 AM Mykyta Rotskyi <notifications@github.com>
wrote:

> @sharadjain21 <https://github.com/sharadjain21> - do you want to install
> go 1.16 arm64 on your mac m1 and try a simple go install
> k8s.io/kubernetes/cmd/kubectl?
>
> Why not use binary?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/97743#issuecomment-787504774>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AKACI4ME6OVOJH5E24KV35DTBKICRANCNFSM4VWVF6PQ>
> .
>

Rotzke commented:
> you mean to clone k8s repo and then make it .. right?
> […](#)

curl -LO https://storage.googleapis.com/kubernetes-release/release/v1.20.0/bin/darwin/amd64/kubectl
chmod +x ./kubectl
sudo mv ./kubectl /usr/local/bin/kubectl

sharadjain21 commented:
but this is for amd64 .. would it hold good for arm64 as well.. as i need
for M1.

On Mon, Mar 1, 2021 at 12:43 AM Mykyta Rotskyi <notifications@github.com>
wrote:

> you mean to clone k8s repo and then make it .. right?
> … <#m_-307530042857058300_>
>
> curl -LO
> https://storage.googleapis.com/kubernetes-release/release/v1.20.0/bin/darwin/amd64/kubectl
> chmod +x ./kubectl
> sudo mv ./kubectl /usr/local/bin/kubectl
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/97743#issuecomment-787505653>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AKACI4MSQLDJ4V32RECW2BLTBKIUVANCNFSM4VWVF6PQ>
> .
>

Rotzke commented:
> but this is for amd64 .. would it hold good for arm64 as well.. as i need for M1.
> […](#)

You won't see difference, that is why M1 has Rosetta 2 emulator. Until they roll out bottle for kubectl in brew. Up to you, just wanted to help.

sharadjain21 commented:
meanwhile i tried the same and got other issues.

om@Oms-MacBook-Pro ~ % brew install --build-from-source kubernetes-cli

==> *Downloading
https://homebrew.bintray.com/bottles/go-1.16.arm64_big_sur.bottle.tar.gz
<https://homebrew.bintray.com/bottles/go-1.16.arm64_big_sur.bottle.tar.gz>*

==> *Downloading from
https://d29vzk4ow07wi7.cloudfront.net/c1f259427b953cb5564af663538a696bb79820f1b23b7baa940ab2c4e21bc684?response-content-disposition=attachment%3Bfilename%3D%22go-1.16.arm64_big_sur.bo
<https://d29vzk4ow07wi7.cloudfront.net/c1f259427b953cb5564af663538a696bb79820f1b23b7baa940ab2c4e21bc684?response-content-disposition=attachment%3Bfilename%3D%22go-1.16.arm64_big_sur.bo>*

########################################################################
100.0%

==> *Cloning https://github.com/kubernetes/kubernetes.git
<https://github.com/kubernetes/kubernetes.git>*

Cloning into '/Users/om/Library/Caches/Homebrew/kubernetes-cli--git'...

Updating files: 100% (22733/22733), done.

==> *Checking out tag v1.20.4*

HEAD is now at e87da0bd Release commit for Kubernetes v1.20.4

==> *Installing dependencies for kubernetes-cli: **go*

==> *Installing kubernetes-cli dependency: **go*

==> *Pouring go-1.16.arm64_big_sur.bottle.tar.gz*

/opt/homebrew/Library/Homebrew/download_strategy.rb:94: warning:
conflicting chdir during another chdir block

🍺  /opt/homebrew/Cellar/go/1.16: 9,967 files, 504.8MB

==> *Installing **kubernetes-cli*

==> *make WHAT=cmd/kubectl*

Last 15 lines from /Users/om/Library/Logs/Homebrew/kubernetes-cli/01.make:

    ./vendor/k8s.io/code-generator/cmd/conversion-gen

touch: _output/bin/go-bindata: No such file or directory

make[1]: *** [_output/bin/go-bindata] Error 1

make[1]: *** Waiting for unfinished jobs....

touch: _output/bin/prerelease-lifecycle-gen: No such file or directory

make[1]: *** [_output/bin/prerelease-lifecycle-gen] Error 1

touch: _output/bin/deepcopy-gen: No such file or directory

make[1]: *** [_output/bin/deepcopy-gen] Error 1

touch: _output/bin/defaulter-gen: No such file or directory

make[1]: *** [_output/bin/defaulter-gen] Error 1

touch: _output/bin/conversion-gen: No such file or directory

make[1]: *** [_output/bin/conversion-gen] Error 1

touch: _output/bin/openapi-gen: No such file or directory

make[1]: *** [_output/bin/openapi-gen] Error 1

make: *** [generated_files] Error 2


Do not report this issue to Homebrew/brew or Homebrew/core!


These open issues may also help:

kubernetes-cli: Add support for darwin/arm64
https://github.com/Homebrew/homebrew-core/pull/69931

kubernetes-cli: build for Apple Silicon
https://github.com/Homebrew/homebrew-core/pull/68455



Thanks

Sharad

On Mon, Mar 1, 2021 at 12:29 AM Davanum Srinivas <notifications@github.com>
wrote:

> @sharadjain21 <https://github.com/sharadjain21> - do you want to install
> go 1.16 arm64 on your mac m1 and try a simple go install
> k8s.io/kubernetes/cmd/kubectl?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/97743#issuecomment-787502921>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AKACI4KT6FTJHRIDAPXA4M3TBKHCHANCNFSM4VWVF6PQ>
> .
>

sharadjain21 commented:
cool.. thanks i will try that way .. goal is to be able to use k8s :)

On Mon, Mar 1, 2021 at 12:49 AM Mykyta Rotskyi <notifications@github.com>
wrote:

> but this is for amd64 .. would it hold good for arm64 as well.. as i need
> for M1.
> … <#m_-3522597250455255052_>
>
> You won't see difference, that is why M1 has Rosetta 2 emulator. Until
> they roll out bottle for kubectl in brew. Up to you, just wanted to help.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/97743#issuecomment-787506688>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AKACI4M4WAZYMNOY745RL6TTBKJLTANCNFSM4VWVF6PQ>
> .
>

liggitt commented:
should be able to rebase this now and use go1.16 bits

dims commented:
testing using latest master + this PR on my mac m1:
```
build/run.sh make generated_files && make quick-release-images
```

dims commented:
@BenTheElder can you verify as well?

/assign @BenTheElder

dims commented:
/hold cancel

bl-ue commented:
/test pull-kubernetes-verify

bl-ue commented:
Odd...it's a shellcheck issue 🤔

BenTheElder commented on hack/lib/golang.sh:
```diff
@@ -206,25 +208,29 @@ kube::golang::setup_platforms() {
   elif [[ "${KUBE_FASTBUILD:-}" == "true" ]]; then
-    KUBE_SERVER_PLATFORMS=(linux/amd64)
+    host_arch=$(kube::util::host_arch)
+    if [[ "${host_arch}" != "amd64" && "${host_arch}" != "arm64" ]]; then
```
can we add a comment justifying this?

dims replied:
Done! thanks.

BenTheElder replied:
er, but why not do this for all platforms? I think that's still missing.

BenTheElder replied:
Ping 🙃

dims replied:
@BenTheElder limited by `kube-cross`, let's roll this out for arm64 (linux + darwin) and then we can hit the other arch(es)

BenTheElder replied:
ACK.

BenTheElder replied:
we should probably also update the makefile docs / help text, but that can be a follow up

BenTheElder commented:
I can continue to review the build scripts / typecheck, but I actually do not have access to M1 hardware. Paging 🍎 cc @munnerz @mauilion

dims commented:
@BenTheElder dang! i thought you ordered one :)

justaugustus commented:
For RelEng:
/lgtm
/approve
cc: @kubernetes/release-engineering

alenkacz commented:
@dims done and passed. Let me know if you need the logs :)

dims commented:
@alenkacz thanks for the quick response! 

@BenTheElder @cblecker this is ready to merge (and there's just the 2 of you in test/typecheck/OWNERS as approvers!)

thanks,
dims

pires commented:
Does this require Docker for Mac? I mean, I'm assuming in order to properly test this a VM, eg minikube, won't cut it.

ghuntley commented:
:wave: I'm building this PR from a brand new Mac M1. I'll update this comment when complete with a ✅ .

So far to build I've had to do the following on a fresh M1:

- install bash via brew
- install gnu-tar via brew

alenkacz commented:
yeah I had to download those two deps as well via brew. Also yes, I am using docker for mac preview (for M1). There's really no other choice right now :)

pires commented:
Seems Docker for Mac TP M1 on my Mini on 11.2.2 is not working. If this is still open when I get Docker to work, I'll share my feedback.

endocrimes commented:
Everything succeeds for me on an M1 MBP running 11.2

/lgtm

`docker version` output:

```
Client: Docker Engine - Community
 Cloud integration: 1.0.8
 Version:           20.10.3
 API version:       1.41
 Go version:        go1.13.15
 Git commit:        48d30b5
 Built:             Fri Jan 29 14:28:27 2021
 OS/Arch:           darwin/amd64 (rosetta)
 Context:           default
 Experimental:      true

Server: Docker Engine - Community
 Engine:
  Version:          20.10.3
  API version:      1.41 (minimum version 1.12)
  Go version:       go1.13.15
  Git commit:       46229ca
  Built:            Fri Jan 29 14:32:07 2021
  OS/Arch:          linux/arm64
  Experimental:     false
 containerd:
  Version:          1.4.3
  GitCommit:        269548fa27e0089a8b8278fc4fc781d7f65a939b
 runc:
  Version:          1.0.0-rc92
  GitCommit:        ff819c7e9184c13b7c2607fe6c30ae19403a7aff
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

endocrimes commented:
For good measure, I'll also include the output from `make all` here:

```
[kubernetes(dims-m1)] $ time make all
+++ [0303 14:06:06] Building go targets for darwin/arm64:
    cmd/kube-proxy
    cmd/kube-apiserver
    cmd/kube-controller-manager
    cmd/kubelet
    cmd/kubeadm
    cmd/kube-scheduler
    vendor/k8s.io/kube-aggregator
    vendor/k8s.io/apiextensions-apiserver
    cluster/gce/gci/mounter
    cmd/kubectl
    cmd/kubectl-convert
    cmd/gendocs
    cmd/genkubedocs
    cmd/genman
    cmd/genyaml
    cmd/genswaggertypedocs
    cmd/linkcheck
    vendor/github.com/onsi/ginkgo/ginkgo
    test/e2e/e2e.test
    cluster/images/conformance/go-runner
    cmd/kubemark
    vendor/github.com/onsi/ginkgo/ginkgo

make all  289.86s user 44.61s system 552% cpu 1:00.56 total
```

dims commented:
@endocrimes thanks! would you be able to verify that say `kubectl` works without rosetta2?

pires commented:
Passed on Docker for Mac 3.1.0, Mini M1 11.2.2 🎉 

<details>
  <summary>Go and Docker versions</summary>

```
$ go version
go version go1.16 darwin/amd64

$ docker version
Client: Docker Engine - Community
 Cloud integration: 1.0.8
 Version:           20.10.3
 API version:       1.41
 Go version:        go1.13.15
 Git commit:        48d30b5
 Built:             Fri Jan 29 14:28:27 2021
 OS/Arch:           darwin/amd64 (rosetta)
 Context:           default
 Experimental:      true

Server: Docker Engine - Community
 Engine:
  Version:          20.10.3
  API version:      1.41 (minimum version 1.12)
  Go version:       go1.13.15
  Git commit:       46229ca
  Built:            Fri Jan 29 14:32:07 2021
  OS/Arch:          linux/arm64
  Experimental:     false
 containerd:
  Version:          1.4.3
  GitCommit:        269548fa27e0089a8b8278fc4fc781d7f65a939b
 runc:
  Version:          1.0.0-rc92
  GitCommit:        ff819c7e9184c13b7c2607fe6c30ae19403a7aff
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

</details>

endocrimes commented:
@dims here you go!

```
[kubernetes(dims-m1)] $ ./_output/bin/kubectl --version
Error: unknown flag: --version
See 'kubectl --help' for usage.
[kubernetes(dims-m1)] $ ./_output/bin/kubectl version  
Client Version: version.Info{Major:"1", Minor:"21+", GitVersion:"v1.21.0-beta.0.254+f719624654c68d", GitCommit:"f719624654c68d143cad4e4af2f21cca0bfde8fd", GitTreeState:"clean", BuildDate:"2021-03-03T13:06:06Z", GoVersion:"go1.16", Compiler:"gc", Platform:"darwin/arm64"}
Server Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.4+k3s-8f9e5628-dirty", GitCommit:"8f9e5628c1b5f0a673cde6051cea1afb65dcff7f", GitTreeState:"dirty", BuildDate:"2020-12-05T00:04:10Z", GoVersion:"go1.16", Compiler:"gc", Platform:"linux/amd64"}
[kubernetes(dims-m1)] $ ./_output/bin/kubectl get pods
NAME                                   READY   STATUS    RESTARTS   AGE
svclb-home-assistant-j6gw8             1/1     Running   1          4d14h
svclb-plex-jk9p7                       1/1     Running   1          4d14h
svclb-blocky-udp-9l5br                 1/1     Running   1          4d14h
svclb-blocky-tcp-rxgth                 1/1     Running   1          4d14h
home-assistant-postgresql-0            1/1     Running   1          4d14h
athens-athens-proxy-6996f79987-9qs2j   1/1     Running   1          4d14h
photoprism-69d744dc9c-z2tr4            1/1     Running   1          4d14h
blocky-6c8f4c5fd-qkqm6                 1/1     Running   1          4d14h
home-assistant-66f459b4d7-c9sjb        1/1     Running   1          4d14h
```

dims commented:
@endocrimes thanks!

/meowvie thanks

stevesloka commented:
Worked fine for me! 🎉

erismaster commented:
Hello! I am from the Bug Triage team! Friendly reminder that code freeze is starting March 9th, 2021 (about 1 week from now). We want to ensure that each PR has a chance to be merged before code freeze. This looks like its close to being approved and merged but still commenting for awareness 😄

estesp commented:
Just FYI--I got my M1 before the end of 2020 and had a fairly hacked up environment (brew under Rosetta and hadn't updated to the M1-enabled `/opt/homebrew` configuration) as well as my iTerm started with arch x86_64 to "force" Rosetta.

To get the changes in this PR to work I had to undo my arch hackiness (which I've needed to for awhile, so thanks for forcing my hand @dims! 😂 ) and make sure I was running fully under `linux/arm64` mode with Go + brew + terminal bash all aligned. Once I did that everything worked properly. `file kubectl` shows it building as a proper `aarch64` binary.
```
$ file kubectl
kubectl: Mach-O 64-bit executable arm64
```

dims commented:
woohoo @estesp !

BenTheElder commented:
@dims see: https://github.com/kubernetes/kubernetes/pull/97743#discussion_r586543400

dims commented:
@BenTheElder sorry missed it. responded there

BenTheElder commented:
/lgtm
/approve
/triage accepted
/test all

BenTheElder commented:
/test pull-kubernetes-cross

justaugustus commented:
/test pull-kubernetes-e2e-kind

endocrimes commented:
:sparkles: :rocket: :sparkling_heart:

dims commented:
Thanks everyone for helping with testing this! 

@endocrimes @estesp @stevesloka  @pires @alenkacz  @ghuntley  (apologies if i missed anyone!)

![image](https://user-images.githubusercontent.com/23304/109988593-aef37080-7cd5-11eb-8862-490f7d87d5c1.png)

timkuilman commented:
This was super interesting to follow. I managed to build kubectl from source and it works like a charm 🔥

dims commented:
for those who end up coming here ... if you need just kubectl ... you don't need to build the full release. just do:

```
make kubectl
```
